### PR TITLE
Move hearbeat to use android worker

### PR DESCRIPTION
**Pull request by ShivamPokhriyal**
Product Note: Improved heartbeat periodicity and reliability.

### CODE REVIEW

**Comment by dimagimon** [comment]
[32% (0.0%) vs master 32%](https://jenkins.dimagi.com/job/commcare-android-pr-job-builder/462215/)

File: app/src/org/commcare/heartbeat/HeartbeatWorker.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+class HeartbeatWorker(context: Context, workerParams: WorkerParameters):
+        Worker(context, workerParams) {
+    private val requester = CommCareApplication.instance().heartbeatRequester
+    override fun doWork(): Result {
+        try {
+            if (!CommCareApplication.instance().session.heartbeatSucceededForSession()) {
+                requester.makeRequest()
+            }
+        } catch (e: Exception) {
```
Comment (shubham1g5) [inline app/src/org/commcare/heartbeat/HeartbeatWorker.kt:21]: We should handle different network failures here like we do for other network processes like sync and updates and use Result.retry() for temporary failures.

File: app/src/org/commcare/heartbeat/HeartbeatWorker.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * @author $|-|!˅@M
+ */
+class HeartbeatWorker(context: Context, workerParams: WorkerParameters):
+        Worker(context, workerParams) {
+    private val requester = CommCareApplication.instance().heartbeatRequester
+    override fun doWork(): Result {
+        try {
+            if (!CommCareApplication.instance().session.heartbeatSucceededForSession()) {
```
Comment (shubham1g5) [inline app/src/org/commcare/heartbeat/HeartbeatWorker.kt:18]: we should also have a `session.isActive` check here to make sure we don't fire a heartbeat after the session has expired.

**Review by shubham1g5 — CHANGES_REQUESTED** [review]

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please

**Comment by dimagimon** [comment]
[32% (0.0%) vs master 32%](https://jenkins.dimagi.com/job/commcare-android-pr-job-builder/462221/)

**Review by shubham1g5 — APPROVED** [review]

**Pull request merged**